Source/d_loop.c

```diff
@@ -0,0 +1,864 @@
+
+    //!
+    // @category net
+    // @arg <n>
+    //
+    // Send n extra tics in every packet as insurance against dropped
+    // packets.
+    //
+
+    i = M_CheckParmWithArgs("-extratics", 1);
+
+    if (i > 0)
+        settings->extratics = atoi(myargv[i+1]);
+    else
+        settings->extratics = 1;
+
+    //!
+    // @category net
+    // @arg <n>
+    //
+    // Reduce the resolution of the game by a factor of n, reducing
+    // the amount of network bandwidth needed.
+    //
+
+    i = M_CheckParmWithArgs("-dup", 1);
+
+    if (i > 0)
+        settings->ticdup = atoi(myargv[i+1]);
+    else
+        settings->ticdup = 1;
+
+    if (net_client_connected)
+    {
+        // Send our game settings and block until game start is received
+        // from the server.
+
+        NET_CL_StartGame(settings);
+        BlockUntilStart(settings, callback);
+
+        // Read the game settings that were received.
+
+        NET_CL_GetSettings(settings);
+    }
+
+    if (drone)
+    {
+        settings->consoleplayer = 0;
+    }
+
+    // Set the local player and playeringame[] values.
+
+    localplayer = settings->consoleplayer;
+
+    for (i = 0; i < NET_MAXPLAYERS; ++i)
+    {
+        local_playeringame[i] = i < settings->num_players;
+    }
+
+    // Copy settings to global variables.
+
+    ticdup = settings->ticdup;
+    new_sync = settings->new_sync;
+
+    if (ticdup < 1)
+    {
+        I_Error("D_StartNetGame: invalid ticdup value (%d)", ticdup);
+    }
+
+    // TODO: Message disabled until we fix new_sync.
+    //if (!new_sync)
+    //{
+    //    printf("Syncing netgames like Vanilla Doom.\n");
+    //}
+}
+
+static int expected_nodes;
+
+static void NET_WaitForLaunch()
+{
+    int i;
+
+    //!
+    // @arg <n>
+    // @category net
+    //
+    // Autostart the netgame when n nodes (clients) have joined the server.
+    //
+
+    i = M_CheckParmWithArgs("-nodes", 1);
+    if (i > 0)
+    {
+        expected_nodes = atoi(myargv[i + 1]);
+    }
+
+    while (net_waiting_for_launch)
+    {
+        int nodes;
+        boolean added;
+
+        if (net_client_received_wait_data
+         && net_client_wait_data.is_controller
+         && expected_nodes > 0)
+        {
+            nodes = net_client_wait_data.num_players
+                  + net_client_wait_data.num_drones;
+
+            if (nodes >= expected_nodes)
+            {
+                NET_CL_LaunchGame();
+                expected_nodes = 0;
+            }
+        }
+
+        NET_Query_CheckAddedToMaster(&added);
+
+        NET_CL_Run();
+        NET_SV_Run();
+
+        if (!net_client_connected)
+        {
+            I_Error("NET_WaitForLaunch: Lost connection to server");
+        }
+
+        I_Sleep(100);
+    }
+}
+
+boolean D_InitNetGame(net_connect_data_t *connect_data)
+{
+    boolean result = false;
+    net_addr_t *addr = NULL;
+    int i;
+
+    // Call D_QuitNetGame on exit:
+
+    //I_AtExit(D_QuitNetGame, true);
```

Comment:
It seems that the server cannot be properly closed with the standard `atexit()`. Should we implement `I_AtExit()`?

Reply:
> Should we implement I_AtExit()?

We probably should (using the implementation in Pr+). But that's orthogonal to this PR. Let's not mix things up too much.